### ISSUE TITLE
GradientState being in sync with the DataLoader by default could be specified in the doc.

### ISSUE THREAD

**Issue by GabPrato** [huggingface/accelerate#1847]
Maybe add a mention on [this page](https://huggingface.co/docs/accelerate/usage_guides/gradient_accumulation) of the doc that the GradientState is by default in sync with the DataLoader. This can have a nasty effect if you don't know about this. When the DataLoader reaches its last batch, the variable `end_of_dataloader` is set to True and Accelerator will update the weights of the model when one calls `optimizer.step()`, even if the number of gradient accumulation steps hasn't been reached. So if you just updated the weights in the penultimate batch after say 500 gradient step accumulations, then you reach the last batch, you only accumulate gradient once and take a step. Even worse, if you do something like:
```python
for batch in dataloader:
    for i in range(0, batch.shape[0], args.minibatch_size):
        with accelerate.accumulate(model):
            output = model(batch[i : i+args.minibatch_size])
            loss = output.loss
            accelerate.backward(loss)
            optimizer.step()
            optimizer.zero_grad()
```
All minibatches in the last batch will update your weights, instead of accumulating. This isn't a bug, because you can turn the feature off with [GradientAccumulationPlugin](https://huggingface.co/docs/accelerate/v0.21.0/en/package_reference/accelerator#accelerate.utils.GradientAccumulationPlugin), but maybe clarify the default behavior in the doc.

**Comment by sgugger** [comment]
cc @muellerzr
